infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -573,73 +563,185 @@ private LogData getLogData(LogEntry entry) {
+     * partially written.
+     *
+     * @param ch
+     * @return
+     * @throws IOException
+     */
+    private LogHeader parseHeader(FileChannel ch) throws IOException {
+        Metadata metadata = parseMetadata(ch);
+        if (metadata == null) {
+            // Partial write on the metadata for the header
+            // Rewind the channel position to the begining of the file
+            ch.position(0);
+            return null;
+        }
 
-            try {
-                Metadata metadata = Metadata.parseFrom(metadataBuf);
+        ByteBuffer buffer = getPayloadForMetadata(ch, metadata);
+        if (buffer == null) {
+            // partial write on the header payload
+            // Rewind the channel position to the begining of the file
+            ch.position(0);
+            return null;
+        }
 
-                byte[] logEntryBuf = new byte[metadata.getLength()];
+        if (getChecksum(buffer.array()) != metadata.getPayloadChecksum()) {
+            throw new DataCorruptionException();
+        }
 
-                o.get(logEntryBuf);
+        LogHeader header;
 
-                LogEntry entry = LogEntry.parseFrom(logEntryBuf);
+        try {
+            header = LogHeader.parseFrom(buffer.array());
+        } catch (InvalidProtocolBufferException e) {
+            throw new DataCorruptionException();
+        }
 
-                if (!noVerify) {
-                    if (metadata.getChecksum() != getChecksum(entry.toByteArray())) {
-                        log.error("Checksum mismatch detected while trying to read file {}",
-                                sh.fileName);
-                        throw new DataCorruptionException();
-                    }
-                }
+        return header;
+    }
 
-                sh.getKnownAddresses().put(entry.getGlobalAddress(),
-                        new AddressMetaData(metadata.getChecksum(),
-                                metadata.getLength(), channelOffset));
+    /**
+     * Parse an entry.
+     *
+     * @param ch
+     * @param metadata
+     * @return
+     * @throws IOException
+     */
+    private LogEntry parseEntry(FileChannel ch, Metadata metadata) throws IOException {
+        if (metadata == null) {
+            // The metadata for this entry was partial written
+            return null;
+        }
 
-                channelOffset += metadata.getLength();
+        ByteBuffer buffer = getPayloadForMetadata(ch, metadata);
+        if (buffer == null) {
+            // partial write on the entry
+            // rewind the channel position to point before
+            // the metadata field for this partially written payload
+            ch.position(ch.position() - METADATA_SIZE);
+            return null;
+        }
 
-            } catch (InvalidProtocolBufferException e) {
+        if (!noVerify) {
+            if (metadata.getPayloadChecksum() != getChecksum(buffer.array())) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)